File: Core/src/main/java/dev/lynith/core/ui/components/features/Children.java
package dev.lynith.core.ui.components.features;

import dev.lynith.core.ui.components.Component;

import java.util.List;

public interface Children {

    List<Component<?>> getChildren();

    default void addChild(Component<?> child) {
        getChildren().add(child);
    }

    default void addChildren(Component<?>... children) {
        for (Component<?> child : children) {
            addChild(child);
        }
    }

    default void addChildren(List<Component<?>> children) {
        getChildren().addAll(children);
    }

    default void removeChild(Component<?> child) {
        getChildren().remove(child);
    }

    default void setChildren(List<Component<?>> children) {
        getChildren().clear();
        getChildren().addAll(children);
    }

    default void clearChildren() {
        getChildren().clear();
    }

}

File: Core/src/main/kotlin/dev/lynith/core/ui/components/features/Children.kt
package dev.lynith.core.ui.components.features

import dev.lynith.core.ui.components.Component

interface Children {

    val children: MutableList<Component<*, *>>

    fun child(child: Component<*, *>) {
        children.add(child)
    }

    fun children(vararg children: Component<*, *>) {
        for (c in children) {
            child(c.apply {
                parent = this
            })
        }
    }

    fun children(children: List<Component<*, *>>) {
        this.children.addAll(children)
    }

    fun removeChild(child: Component<*, *>) {
        children.remove(child)
    }

    fun clearChildren() {
        children.clear()
    }
}
